### PR TITLE
ENG-11769: Update Generic SAML Draft to return SP Metadata

### PR DESCRIPTION
**Pull request by gabfelp**
## Description of the change

This PR updates the Generic SAML Draft terraform resource (and the dataSource) to return the fields recently added in the API for the Entity ID, Both ACS URLs with their respective index and the SLO URL.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing

First, I updated my env variables to use valid  `CYRAL_TF_CONTROL_PLANE`, `CYRAL_TF_CLIENT_ID` and `CYRAL_TF_CLIENT_SECRET`

I then run the `make` command, which also runs the tests and everything ran fine:
```
go clean -i github.com/cyralinc/terraform-provider-cyral/...
rm -f terraform-provider-cyral_v4.2.0
rm -rf ./out
rm -rf ~/.terraform.d/plugins/local/terraform/cyral/4.2.0
gofmt -w .
mkdir -p out/
# Build for both MacOS and Linux
GOOS=darwin GOARCH=amd64 go build -o out/darwin_amd64/terraform-provider-cyral_v4.2.0 .
GOOS=darwin GOARCH=arm64 go build -o out/darwin_arm64/terraform-provider-cyral_v4.2.0 .
GOOS=linux GOARCH=amd64 go build -o out/linux_amd64/terraform-provider-cyral_v4.2.0 .
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_amd64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_arm64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/linux_amd64
cp out/darwin_arm64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_arm64
cp out/linux_amd64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/linux_amd64
cp out/darwin_amd64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_amd64
go test github.com/cyralinc/terraform-provider-cyral/... -v -race -timeout 20m
?       github.com/cyralinc/terraform-provider-cyral    [no test files]
=== RUN   TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue
2023/05/18 14:36:07 [DEBUG] Init NewClient
2023/05/18 14:36:07 [DEBUG] TokenSource: &{0xc000012e88 {0 0} <nil>}
2023/05/18 14:36:07 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue (0.00s)
=== RUN   TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse
2023/05/18 14:36:07 [DEBUG] Init NewClient
2023/05/18 14:36:07 [DEBUG] TokenSource: &{0xc000012eb8 {0 0} <nil>}
2023/05/18 14:36:07 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse (0.00s)
=== RUN   TestNewClient_WhenClientIDIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientIDIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenClientSecretIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientSecretIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/client     (cached)
=== RUN   TestAccDatalabelDataSource
=== PAUSE TestAccDatalabelDataSource
=== RUN   TestAccIntegrationIdPSAMLDataSource
=== PAUSE TestAccIntegrationIdPSAMLDataSource
=== RUN   TestAccRepositoryDataSource
=== PAUSE TestAccRepositoryDataSource
=== RUN   TestAccRoleDataSource
=== PAUSE TestAccRoleDataSource
=== RUN   TestAccSAMLCertificateDataSource
=== PAUSE TestAccSAMLCertificateDataSource
=== RUN   TestAccSAMLConfigurationDataSource
=== PAUSE TestAccSAMLConfigurationDataSource
=== RUN   TestAccSidecarBoundPortsDataSource
=== PAUSE TestAccSidecarBoundPortsDataSource
=== RUN   TestAccSidecarCftTemplateDataSource
=== PAUSE TestAccSidecarCftTemplateDataSource
=== RUN   TestAccSidecarIDDataSource
=== PAUSE TestAccSidecarIDDataSource
=== RUN   TestAccSidecarInstanceIDsDataSource
=== PAUSE TestAccSidecarInstanceIDsDataSource
=== RUN   TestIntegrationsData_GetValue_Default
--- PASS: TestIntegrationsData_GetValue_Default (0.00s)
=== RUN   TestIntegrationsData_GetValue_Splunk
--- PASS: TestIntegrationsData_GetValue_Splunk (0.00s)
=== RUN   TestAccProvider
--- PASS: TestAccProvider (0.00s)
=== RUN   TestAccDatalabelResource
=== PAUSE TestAccDatalabelResource
=== RUN   TestAccDatadogIntegrationResource
=== PAUSE TestAccDatadogIntegrationResource
=== RUN   TestAccELKIntegrationResource
=== PAUSE TestAccELKIntegrationResource
=== RUN   TestAccHCVaultIntegrationResource
=== PAUSE TestAccHCVaultIntegrationResource
=== RUN   TestAccIntegrationIdPSAMLDraftResource
=== PAUSE TestAccIntegrationIdPSAMLDraftResource
=== RUN   TestAccIntegrationIdPSAMLResource
=== PAUSE TestAccIntegrationIdPSAMLResource
=== RUN   TestAccIdPIntegrationResource
=== PAUSE TestAccIdPIntegrationResource
=== RUN   TestAccLogstashIntegrationResource
=== PAUSE TestAccLogstashIntegrationResource
=== RUN   TestAccLookerIntegrationResource
=== PAUSE TestAccLookerIntegrationResource
=== RUN   TestAccDuoMFAIntegrationResource
=== PAUSE TestAccDuoMFAIntegrationResource
=== RUN   TestAccPagerDutyIntegrationResource
=== PAUSE TestAccPagerDutyIntegrationResource
=== RUN   TestAccSlackAlertsIntegrationResource
=== PAUSE TestAccSlackAlertsIntegrationResource
=== RUN   TestAccSplunkIntegrationResource
=== PAUSE TestAccSplunkIntegrationResource
=== RUN   TestAccSumoLogicIntegrationResource
=== PAUSE TestAccSumoLogicIntegrationResource
=== RUN   TestAccMsTeamsIntegrationResource
=== PAUSE TestAccMsTeamsIntegrationResource
=== RUN   TestAccPolicyRuleResource
=== PAUSE TestAccPolicyRuleResource
=== RUN   TestPolicyRuleResourceUpgradeV0
--- PASS: TestPolicyRuleResourceUpgradeV0 (0.00s)
=== RUN   TestAccPolicyResource
=== PAUSE TestAccPolicyResource
=== RUN   TestAccRepositoryAccessGatewayResource
=== PAUSE TestAccRepositoryAccessGatewayResource
=== RUN   TestAccRepositoryAccessRulesResource
=== PAUSE TestAccRepositoryAccessRulesResource
=== RUN   TestAccRepositoryBindingResource
=== PAUSE TestAccRepositoryBindingResource
=== RUN   TestAccRepositoryConfAnalysisResource
=== PAUSE TestAccRepositoryConfAnalysisResource
=== RUN   TestRepositoryConfAnalysisResourceUpgradeV0
--- PASS: TestRepositoryConfAnalysisResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryConfAuthResource
=== PAUSE TestAccRepositoryConfAuthResource
=== RUN   TestRepositoryConfAuthResourceUpgradeV0
--- PASS: TestRepositoryConfAuthResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryDatamapResource
=== PAUSE TestAccRepositoryDatamapResource
=== RUN   TestAccRepositoryNetworkAccessPolicyResource
=== PAUSE TestAccRepositoryNetworkAccessPolicyResource
=== RUN   TestAccRepositoryResource
=== PAUSE TestAccRepositoryResource
=== RUN   TestAccRepositoryUserAccountResource
=== PAUSE TestAccRepositoryUserAccountResource
=== RUN   TestAccRoleSSOGroupsResource
=== PAUSE TestAccRoleSSOGroupsResource
=== RUN   TestRoleSSOGroupsResourceUpgradeV0
--- PASS: TestRoleSSOGroupsResourceUpgradeV0 (0.00s)
=== RUN   TestAccRoleResource
=== PAUSE TestAccRoleResource
=== RUN   TestAccSidecarCredentialsResource
=== PAUSE TestAccSidecarCredentialsResource
=== RUN   TestSidecarListenerResource
=== PAUSE TestSidecarListenerResource
=== RUN   TestAccSidecarResource
=== PAUSE TestAccSidecarResource
=== RUN   TestElementsMatch
--- PASS: TestElementsMatch (0.00s)
=== CONT  TestAccIntegrationIdPSAMLDataSource
=== CONT  TestAccSplunkIntegrationResource
=== CONT  TestAccRepositoryConfAuthResource
=== CONT  TestAccDatadogIntegrationResource
=== CONT  TestAccPagerDutyIntegrationResource
=== CONT  TestAccDuoMFAIntegrationResource
=== CONT  TestAccLogstashIntegrationResource
=== CONT  TestAccIdPIntegrationResource
--- PASS: TestAccSplunkIntegrationResource (9.69s)
=== CONT  TestAccSAMLConfigurationDataSource
=== CONT  TestAccLookerIntegrationResource
--- PASS: TestAccDatadogIntegrationResource (9.76s)
--- PASS: TestAccDuoMFAIntegrationResource (9.81s)
=== CONT  TestAccIntegrationIdPSAMLResource
--- PASS: TestAccPagerDutyIntegrationResource (9.86s)
=== CONT  TestAccIntegrationIdPSAMLDraftResource
=== CONT  TestAccSidecarInstanceIDsDataSource
--- PASS: TestAccLogstashIntegrationResource (17.58s)
=== CONT  TestAccELKIntegrationResource
--- PASS: TestAccLookerIntegrationResource (8.82s)
--- PASS: TestAccRepositoryConfAuthResource (19.01s)
=== CONT  TestAccSidecarCftTemplateDataSource
=== CONT  TestAccHCVaultIntegrationResource
--- PASS: TestAccSAMLConfigurationDataSource (11.58s)
--- PASS: TestAccIntegrationIdPSAMLDataSource (21.45s)
=== CONT  TestAccDatalabelResource
--- PASS: TestAccIntegrationIdPSAMLDraftResource (12.56s)
=== CONT  TestAccSidecarBoundPortsDataSource
=== CONT  TestAccSidecarResource
--- PASS: TestAccSidecarInstanceIDsDataSource (7.51s)
--- PASS: TestAccELKIntegrationResource (8.55s)
=== CONT  TestAccRoleSSOGroupsResource
--- PASS: TestAccHCVaultIntegrationResource (8.22s)
=== CONT  TestAccSidecarCredentialsResource
--- PASS: TestAccDatalabelResource (8.43s)
=== CONT  TestSidecarListenerResource
--- PASS: TestAccSidecarCftTemplateDataSource (14.05s)
=== CONT  TestAccRoleResource
=== CONT  TestAccRepositoryUserAccountResource
--- PASS: TestAccIntegrationIdPSAMLResource (23.60s)
--- PASS: TestAccSidecarBoundPortsDataSource (14.40s)
=== CONT  TestAccRepositoryDatamapResource
=== CONT  TestAccRepositoryResource
--- PASS: TestAccSidecarCredentialsResource (7.37s)
--- PASS: TestAccRoleSSOGroupsResource (13.60s)
=== CONT  TestAccPolicyRuleResource
--- PASS: TestAccIdPIntegrationResource (46.50s)
=== CONT  TestAccRepositoryAccessRulesResource
--- PASS: TestAccRoleResource (17.91s)
=== CONT  TestAccRepositoryAccessGatewayResource
=== CONT  TestAccRepositoryConfAnalysisResource
--- PASS: TestAccSidecarResource (32.63s)
--- PASS: TestAccRepositoryResource (25.97s)
=== CONT  TestAccPolicyResource
--- PASS: TestAccRepositoryDatamapResource (26.92s)
=== CONT  TestAccMsTeamsIntegrationResource
--- PASS: TestAccPolicyRuleResource (23.92s)
=== CONT  TestAccSumoLogicIntegrationResource
=== CONT  TestAccRepositoryBindingResource
--- PASS: TestAccRepositoryAccessRulesResource (19.79s)
--- PASS: TestAccRepositoryConfAnalysisResource (9.09s)
=== CONT  TestAccRepositoryDataSource
=== CONT  TestAccSidecarIDDataSource
--- PASS: TestSidecarListenerResource (39.04s)
=== CONT  TestAccRepositoryNetworkAccessPolicyResource
--- PASS: TestAccMsTeamsIntegrationResource (6.35s)
--- PASS: TestAccPolicyResource (7.27s)
=== CONT  TestAccSlackAlertsIntegrationResource
--- PASS: TestAccSumoLogicIntegrationResource (7.30s)
=== CONT  TestAccSAMLCertificateDataSource
=== CONT  TestAccDatalabelDataSource
--- PASS: TestAccRepositoryAccessGatewayResource (22.75s)
--- PASS: TestAccSidecarIDDataSource (6.21s)
=== CONT  TestAccRoleDataSource
--- PASS: TestAccSAMLCertificateDataSource (4.24s)
--- PASS: TestAccSlackAlertsIntegrationResource (6.86s)
--- PASS: TestAccRepositoryBindingResource (11.11s)
--- PASS: TestAccRoleDataSource (5.08s)
--- PASS: TestAccRepositoryDataSource (14.93s)
--- PASS: TestAccRepositoryUserAccountResource (52.02s)
--- PASS: TestAccDatalabelDataSource (19.48s)
--- PASS: TestAccRepositoryNetworkAccessPolicyResource (24.04s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral      94.190s
```

To validate it, I created a .tf file with this content:

```
resource "cyral_integration_idp_saml_draft" "test_saml" {
  display_name = "example-okta-integration"
  idp_type     = "okta"
  attributes {
    first_name = "some-first-name"
    last_name  = "some-last-name"
    email      = "some-email"
    groups     = "some-group"
  }
  disable_idp_initiated_login = false
}

data "cyral_integration_idp_saml" "test_source" {
  # These are the filters you can apply.
  display_name = ""
  idp_type     = ""
}
```

And connected to a CP using the latest API for generic SAML.

After running `terraform init` and `terraform apply`, I got:

```
data.cyral_integration_idp_saml.test_source: Reading...
data.cyral_integration_idp_saml.test_source: Read complete after 1s [id=XXXXXXXXXX]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # cyral_integration_idp_saml_draft.test_saml will be created
  + resource "cyral_integration_idp_saml_draft" "test_saml" {
      + disable_idp_initiated_login = false
      + display_name                = "example-okta-integration"
      + id                          = (known after apply)
      + idp_type                    = "okta"
      + service_provider_metadata   = (known after apply)
      + sp_metadata                 = (known after apply)

      + attributes {
          + email      = "some-email"
          + first_name = "some-first-name"
          + groups     = "some-group"
          + last_name  = "some-last-name"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cyral_integration_idp_saml_draft.test_saml: Creating...
cyral_integration_idp_saml_draft.test_saml: Creation complete after 2s [id=XXXXXXXXXX]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

By doing `terraform show`:

```
resource "cyral_integration_idp_saml_draft" "test_saml" {
    disable_idp_initiated_login = false
    display_name                = "example-okta-integration"
    id                          = "ID_INTEGRATION"
    idp_type                    = "okta"
    service_provider_metadata   = [
        {
            assertion_consumer_services = [
                {
                    index = 0
                    url   = "https://CP/auth/realms/default/broker/ID_INTEGRATION/endpoint/clients/ID_INTEGRATION-client"
                },
                {
                    index = 1
                    url   = "https://CP/auth/realms/default/broker/ID_INTEGRATION/endpoint"
                },
            ]
            entity_id                   = "https://CP/auth/realms/default"
            single_logout_url           = "https://CP/auth/realms/default/broker/ID_INTEGRATION/endpoint"
            url                         = "https://CP/v1/integrations/sp-metadata/ID_INTEGRATION"
            xml_document                = "<SP Metadata XML>"
        },
    ]
    sp_metadata                 = <SP Metadata XML> // this was maintained for avoid regression

    attributes {
        email      = "some-email"
        first_name = "some-first-name"
        groups     = "some-group"
        last_name  = "some-last-name"
    }
}
```

Before completing integration, changing any variable in the resource, for instance:
```
resource "cyral_integration_idp_saml_draft" "test_saml" {
  display_name = "changed-name-2"
  idp_type     = "okta"
  attributes {
    first_name = "another"
    last_name  = "another-last-name"
    email      = "any-email"
    groups     = "any-group"
  }
  disable_idp_initiated_login = false
}
```
and doing `terraform apply` will correctly delete and create a new one

After completing integration, doing `terraform apply` and `terraform show`, I was able to see the data_source :

```
data "cyral_integration_idp_saml" "test_source" {
    id       = "XXXXXXX"
    idp_list = [
        {
            attributes     = [
                {
                    email      = "some-email"
                    first_name = "some-first-name"
                    groups     = "some-group"
                    last_name  = "some-last-name"
                },
            ]
            disabled       = false
            display_name   = "example-okta-integration"
            id             = "ID_INTEGRATION"
            idp_descriptor = [
                {
                    disable_force_authentication = true
                    signing_certificate          = "XXX"
                    single_logout_service_url    = "URL"
                    single_sign_on_service_url   = "URL"
                },
            ]
            idp_type       = "okta"
            sp_metadata    = [
                {
                    assertion_consumer_services = [
                        {
                            index = 0
                            url   = "URL"
                        },
                        {
                            index = 1
                            url   = "URL"
                        },
                    ]
                    entity_id                   = "URL"
                    single_logout_url           = "URL"
                    url                         = "URL"
                    xml_document                = "XML"
                },
            ]
        },
    ]
}
```

### CODE REVIEW

**Comment by gabfelp** [comment]
The SonarCloud is failing due to the code replication in defining schemes (data source for listing Generic SAML and resource for Generic SAML 'crud'). However,  we usually go with Schemes this way, so I don't know the best approach here.

FYI @wcmjunior / @jrich523

File: cyral/model_integration_idp_generic_saml.go [hunk 1]
```diff
@@ -62,6 +73,30 @@ func (uatt *RequiredUserAttributes) WriteToSchema(d *schema.ResourceData) error
 	return d.Set("attributes", attributes)
 }
 
+func (spmetadataObj *GenericSAMLSPMetadata) WriteToSchema(d *schema.ResourceData) error {
```
Comment (VictorGFM) [inline cyral/model_integration_idp_generic_saml.go:76]:
```suggestion
func (spMetadataObj *GenericSAMLSPMetadata) WriteToSchema(d *schema.ResourceData) error {
```

File: cyral/resource_cyral_integration_idp_saml_draft_test.go [hunk 2]
```diff
@@ -122,6 +122,28 @@ func setupIntegrationIdPSAMLDraftTest(t *testing.T, draft *GenericSAMLDraft, res
 		}...)
 	}
 
+	// checking SPMetadata content
+	checkFuncs = append(checkFuncs, []resource.TestCheckFunc{
+		resource.TestMatchResourceAttr(resourceFullName, "service_provider_metadata.0.xml_document",
```
Comment (VictorGFM) [inline cyral/resource_cyral_integration_idp_saml_draft_test.go:127]: If we just want to check if the value was set for the field the [TestCheckResourceAttrSet](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/resource#TestCheckResourceAttrSet) may be more appropriated ([See](https://developer.hashicorp.com/terraform/plugin/sdkv2/testing/acceptance-tests/teststep#builtin-check-functions))

File: cyral/data_source_cyral_integration_idp_saml.go [hunk 3]
```diff
@@ -161,6 +174,40 @@ func dataSourceIntegrationIdPSAML() *schema.Resource {
...
+										Computed:    true,
+									},
+									"single_logout_url": {
+										Description: "The single logout URL defined in the SAML Metadata XML (SL0).",
+										Type:        schema.TypeString,
+										Computed:    true,
+									},
+									"assertion_consumer_services": {
+										Description: "The Assertion Consumer Services defined in the SAML Metadata XML.",
+										Type:        schema.TypeSet,
```
Comment (VictorGFM) [inline cyral/data_source_cyral_integration_idp_saml.go:194]: See my comment about that below.
```suggestion
										Type:        schema.TypeList,
```

File: cyral/data_source_cyral_integration_idp_saml.go [hunk 4]
```diff
@@ -34,8 +34,21 @@ func (resp *ListGenericSAMLIdpsResponse) WriteToSchema(d *schema.ResourceData) e
...
+			}
 			spMetadata = append(spMetadata, map[string]interface{}{
-				"xml_document": idp.SPMetadata.XMLDocument,
+				"xml_document":                idp.SPMetadata.XMLDocument,
+				"url":                         idp.SPMetadata.Url,
+				"entity_id":                   idp.SPMetadata.EntityID,
+				"single_logout_url":           idp.SPMetadata.SingleLogoutURL,
+				"assertion_consumer_services": acs,
 			})
 		}
```
Comment (VictorGFM) [inline cyral/data_source_cyral_integration_idp_saml.go:53]:
```suggestion
if idp.SPMetadata != nil {
  spMetadata = append(spMetadata, idp.SPMetadata.ToMap())
}
```

File: cyral/model_integration_idp_generic_saml.go [hunk 5]
```diff
@@ -62,6 +73,30 @@ func (uatt *RequiredUserAttributes) WriteToSchema(d *schema.ResourceData) error
...
+	var spMetadata []interface{}
+	if spmetadataObj != nil {
```
Comment (VictorGFM) [inline cyral/model_integration_idp_generic_saml.go:78]: If we extract the logic for transforming the SP metadata object into a map to a function (E.g.: `ToMap`) we can reuse this code in the data source as well and get rid of this duplicated code.
```Go
if spMetadataObj != nil {
  spMetadata = append(spMetadata, spMetadataObj.ToMap())
}
```

File: cyral/data_source_cyral_integration_idp_saml_test.go [hunk 6]
```diff
@@ -170,6 +170,30 @@ func testIntegrationIdPSAMLDataSourceChecks(t *testing.T, resName, nameFilter, t
 		resource.TestMatchResourceAttr(dataSourceFullName,
 			"idp_list.0.sp_metadata.0.xml_document", nonEmptyRegex,
 		),
+		resource.TestMatchResourceAttr(dataSourceFullName,
```
Comment (VictorGFM) [inline cyral/data_source_cyral_integration_idp_saml_test.go:173]: Same as my comment below regarding using [TestCheckResourceAttrSet](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/resource#TestCheckResourceAttrSet) instead.

**Review by VictorGFM — COMMENTED** [review]
@gabfelp I have a few comments below, please consider them for both the data source and the resource when applicable.

File: cyral/resource_cyral_integration_idp_saml_draft.go [hunk 7]
```diff
@@ -198,6 +203,55 @@ func resourceIntegrationIdPSAMLDraft() *schema.Resource {
 				Description: "The SP Metadata document describing the Cyral service provider for this integration.",
 				Type:        schema.TypeString,
 				Computed:    true,
+				Deprecated:  "Configure service_provider_metadata.xml_document instead. This will be removed in the next major version of the provider.",
+			},
+			"service_provider_metadata": {
+				Description: "The SP Metadata document describing the Cyral service provider for this integration.",
+				Type:        schema.TypeSet,
+				Computed:    true,
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft.go:211]: In this case, when the field is `Computed`, it's not possible to set `MaxItems` (tested it 👍🏻 )

File: cyral/resource_cyral_integration_idp_saml_draft.go [hunk 8]
```diff
@@ -198,6 +203,55 @@ func resourceIntegrationIdPSAMLDraft() *schema.Resource {
 				Description: "The SP Metadata document describing the Cyral service provider for this integration.",
 				Type:        schema.TypeString,
 				Computed:    true,
+				Deprecated:  "Configure service_provider_metadata.xml_document instead. This will be removed in the next major version of the provider.",
+			},
+			"service_provider_metadata": {
+				Description: "The SP Metadata document describing the Cyral service provider for this integration.",
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft.go]: Addr [ee4bdad](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/ee4bdadf82f568dbee77ed5bd2da3cabd9fd9017)

File: cyral/resource_cyral_integration_idp_saml_draft.go [hunk 9]
```diff
@@ -198,6 +203,55 @@ func resourceIntegrationIdPSAMLDraft() *schema.Resource {
 				Description: "The SP Metadata document describing the Cyral service provider for this integration.",
 				Type:        schema.TypeString,
 				Computed:    true,
+				Deprecated:  "Configure service_provider_metadata.xml_document instead. This will be removed in the next major version of the provider.",
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft.go]: Addr [ee4bdad](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/ee4bdadf82f568dbee77ed5bd2da3cabd9fd9017)

File: cyral/resource_cyral_integration_idp_saml_draft.go [hunk 10]
```diff
@@ -198,6 +203,55 @@ func resourceIntegrationIdPSAMLDraft() *schema.Resource {
...
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"entity_id": {
+							Description: "Entity ID defined in th SAML Metadata XML.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"single_logout_url": {
+							Description: "The single logout URL defined in the SAML Metadata XML (SL0).",
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft.go]: Addr [ee4bdad](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/ee4bdadf82f568dbee77ed5bd2da3cabd9fd9017)

File: cyral/resource_cyral_integration_idp_saml_draft.go [hunk 11]
```diff
@@ -198,6 +203,55 @@ func resourceIntegrationIdPSAMLDraft() *schema.Resource {
...
+							Computed:    true,
+						},
+						"assertion_consumer_services": {
+							Description: "The Assertion Consumer Services defined in the SAML Metadata XML.",
+							Type:        schema.TypeSet,
+							Computed:    true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"url": {
+										Description: "The Assertion Consumer Service URL",
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft.go]: Addr [ee4bdad](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/ee4bdadf82f568dbee77ed5bd2da3cabd9fd9017)

File: cyral/resource_cyral_integration_idp_saml_draft.go [hunk 12]
```diff
@@ -198,6 +203,55 @@ func resourceIntegrationIdPSAMLDraft() *schema.Resource {
...
+							Computed:    true,
+						},
+						"single_logout_url": {
+							Description: "The single logout URL defined in the SAML Metadata XML (SL0).",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"assertion_consumer_services": {
+							Description: "The Assertion Consumer Services defined in the SAML Metadata XML.",
+							Type:        schema.TypeSet,
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft.go]: [637f75e](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/637f75e408adf92993fa7023e968c975b2bfc972)

File: cyral/model_integration_idp_generic_saml.go [hunk 13]
```diff
@@ -62,6 +73,30 @@ func (uatt *RequiredUserAttributes) WriteToSchema(d *schema.ResourceData) error
...
+		acs := []map[string]interface{}{}
+		if spmetadataObj.AssertionConsumerServices != nil {
+			for _, assertionConsumerService := range spmetadataObj.AssertionConsumerServices {
+				acsObject := make(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/model_integration_idp_generic_saml.go]: Addr [7b9f350](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/7b9f3503aa1721a658226ab4e45695663018b2ba)

File: cyral/model_integration_idp_generic_saml.go (hunk 1 above)
Comment (gabfelp) [reply cyral/model_integration_idp_generic_saml.go]: Addr [7b9f350](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/7b9f3503aa1721a658226ab4e45695663018b2ba)

File: cyral/resource_cyral_integration_idp_saml_draft_test.go (hunk 2 above)
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft_test.go]: [637f75e](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/637f75e408adf92993fa7023e968c975b2bfc972) addr

File: cyral/data_source_cyral_integration_idp_saml.go (hunk 3 above)
Comment (gabfelp) [reply cyral/data_source_cyral_integration_idp_saml.go]: addr [637f75e](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/637f75e408adf92993fa7023e968c975b2bfc972)

File: cyral/data_source_cyral_integration_idp_saml_test.go (hunk 6 above)
Comment (gabfelp) [reply cyral/data_source_cyral_integration_idp_saml_test.go]: [6c98809](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/6c9880953be078345e8b73fc3c6013a1e196c8e3) addr

File: cyral/resource_cyral_integration_idp_saml_draft_test.go (hunk 2 above)
Comment (gabfelp) [reply cyral/resource_cyral_integration_idp_saml_draft_test.go]: addr [6c98809](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/6c9880953be078345e8b73fc3c6013a1e196c8e3)

File: cyral/model_integration_idp_generic_saml.go (hunk 5 above)
Comment (gabfelp) [reply cyral/model_integration_idp_generic_saml.go]: Addr [7b9f350](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/7b9f3503aa1721a658226ab4e45695663018b2ba)

File: cyral/data_source_cyral_integration_idp_saml.go (hunk 4 above)
Comment (gabfelp) [reply cyral/data_source_cyral_integration_idp_saml.go]: Addr along with toMap [7b9f350](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/7b9f3503aa1721a658226ab4e45695663018b2ba)

File: cyral/model_integration_idp_generic_saml.go [hunk 14]
```diff
@@ -62,6 +73,30 @@ func (uatt *RequiredUserAttributes) WriteToSchema(d *schema.ResourceData) error
 	return d.Set("attributes", attributes)
 }
 
+func (spmetadataObj *GenericSAMLSPMetadata) WriteToSchema(d *schema.ResourceData) error {
+	var spMetadata []interface{}
+	if spmetadataObj != nil {
+		acs := []map[string]interface{}{}
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/model_integration_idp_generic_saml.go]: Addr along with toMap [7b9f350](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/7b9f3503aa1721a658226ab4e45695663018b2ba)

File: cyral/model_integration_idp_generic_saml.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/model_integration_idp_generic_saml.go]: Thanks, addr: [e4ca35d](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/e4ca35d4f1238f257cafe1dd7c1018723b405846)

**Review by VictorGFM — COMMENTED** [review]
Changes look good @gabfelp, please see my single comment below

File: cyral/model_integration_idp_generic_saml.go [hunk 15]
```diff
@@ -62,6 +73,34 @@ func (uatt *RequiredUserAttributes) WriteToSchema(d *schema.ResourceData) error
 	return d.Set("attributes", attributes)
 }
 
+func (spMetadataObj *GenericSAMLSPMetadata) WriteToSchema(d *schema.ResourceData) error {
+	return d.Set("service_provider_metadata", spMetadataObj.ToMap())
+}
+
+func (spMetadataObj *GenericSAMLSPMetadata) ToMap() []any {
```
[the remark this answers is not in the document]
Comment (gabfelp) [reply cyral/model_integration_idp_generic_saml.go]: Thanks! This makes sense. I changed the name to `ToList` in this case 👍🏻 

[7212158](https://github.com/cyralinc/terraform-provider-cyral/pull/394/commits/7212158e8a280708305355dc0678ecfaba27c40e)

**Review by VictorGFM — APPROVED** [review]
LGTM, thanks for working on this @gabfelp. Could you also please add the Jira ticket number to this PR title before merging?

**Pull request merged**